backend/pkg/modules/internal/differ/spec_diffs.go

```diff
@@ -0,0 +1,277 @@
+package differ
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-openapi/spec"
+	_spec "github.com/openclarity/speculator/pkg/spec"
+	_speculator "github.com/openclarity/speculator/pkg/speculator"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openclarity/apiclarity/api/server/models"
+	"github.com/openclarity/apiclarity/api3/common"
+	"github.com/openclarity/apiclarity/api3/global"
+	"github.com/openclarity/apiclarity/backend/pkg/database"
+	"github.com/openclarity/apiclarity/backend/pkg/modules/internal/core"
+	speculatorutils "github.com/openclarity/apiclarity/backend/pkg/utils/speculator"
+)
+
+const (
+	ModuleName = "differ"
+)
+
+type pluginDiffer struct {
+	httpHandler http.Handler
+
+	apiIDToDiffs map[uint]map[[32]byte]global.Diff
+	totalEvents  int
+
+	accessor core.BackendAccessor
+	sync.RWMutex
+}
+
+//nolint:gochecknoinits // was needed for the module implementation of ApiClarity
+func init() {
+	core.RegisterModule(newDiffer)
+}
+
+//nolint:ireturn,nolintlint // was needed for the module implementation of ApiClarity
+func newDiffer(ctx context.Context, accessor core.BackendAccessor) (core.Module, error) {
+	// Use default values
+	plugin := pluginDiffer{
+		httpHandler:  nil,
+		accessor:     accessor,
+		apiIDToDiffs: make(map[uint]map[[32]byte]global.Diff),
+		totalEvents:  0,
+	}
+
+	// TODO do we want to recover some state here?
+
+	go plugin.StartDiffsSender(ctx)
+
+	return &plugin, nil
+}
+
+func (p *pluginDiffer) Name() string {
+	return ModuleName
+}
+
+func (p *pluginDiffer) EventNotify(ctx context.Context, event *core.Event) {
+	var reconstructedDiff *_spec.APIDiff
+	var providedDiff *_spec.APIDiff
+	var err error
+
+	log.Infof("Got new event notification. event=%+v", event)
+
+	apiEvent := event.APIEvent
+	specKey := _speculator.GetSpecKey(apiEvent.HostSpecName, strconv.Itoa(int(apiEvent.DestinationPort)))
+	speculatorAccessor := p.accessor.GetSpeculatorAccessor()
+
+	if !speculatorAccessor.HasProvidedSpec(specKey) && !speculatorAccessor.HasApprovedSpec(specKey) {
+		log.Debugf("No diffs to calculate")
+		return
+	}
+
+	speculatorTelemetry := speculatorutils.ConvertModelsToSpeculatorTelemetry(event.Telemetry)
+
+	reconstructedDiffType := models.DiffTypeNODIFF
+	providedDiffType := models.DiffTypeNODIFF
+	if speculatorAccessor.HasProvidedSpec(specKey) {
+		// calculate diffs base on the event
+		providedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceProvided)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against provided spec: %v", err)
+			return
+		}
+		if err := setAPIEventProvidedDiff(apiEvent, providedDiff); err != nil {
+			log.Errorf("Failed to set api event provided diff: %v", err)
+			return
+		}
+		providedDiffType = convertAPIDiffType(providedDiff.Type)
+	} else if speculatorAccessor.HasApprovedSpec(specKey) {
```

Comment:
should it be under else? or in a different if? (do we want to calculate diffs from provided AND reconstructed when we have both? or just provided?)

Reply:
should be two different ifs